vktrace/vktrace_common/vktrace_trace_packet_utils.c

```diff
@@ -113,16 +114,163 @@ uint64_t vktrace_get_time() {
 uint64_t vktrace_get_time() { return 0; }
 #endif
 
+uint64_t get_endianess() {
+    uint64_t x = 1;
+    return *((char*)&x) ? VKTRACE_BIG_ENDIAN : VKTRACE_LITTLE_ENDIAN;
+}
+
+uint64_t get_arch() {
+    uint64_t rval;
+#if defined(PLATFORM_LINUX)
+    struct utsname buf;
+    uname(&buf);
+    strncpy((char*)&rval, buf.machine, sizeof(uint64_t));
+#else
+    SYSTEM_INFO systemInfo;
+    char* arch;
+    GetSystemInfo(&systemInfo);
+    switch (systemInfo.wProcessorArchitecture) {
+        case PROCESSOR_ARCHITECTURE_AMD64:
+            arch = "x86_64";
+            break;
+        case PROCESSOR_ARCHITECTURE_ARM:
+            arch = "ARM";
+            break;
+        case PROCESSOR_ARCHITECTURE_IA64:
+            arch = "ia64";
+            break;
+        case PROCESSOR_ARCHITECTURE_INTEL:
+            arch = "x86";
+            break;
+        default:
+            arch = "";
+            break;
+    }
+    strncpy((char*)&rval, arch, sizeof(uint64_t));
+#endif
+    return rval;
+}
+
+uint64_t get_os() {
+    uint64_t rval;
+#if defined(ANDROID)
+    strncpy((char*)&rval, "Android", sizeof(uint64_t));
+#elif defined(PLATFORM_LINUX)
+    strncpy((char*)&rval, "Linux", sizeof(uint64_t));
+#else
+    strncpy((char*)&rval, "Windows", sizeof(uint64_t));
+#endif
+    return rval;
+}
+
+static VkPhysicalDeviceProperties* get_physicalDeviceProperties() {
+    VkPhysicalDeviceProperties* rval = NULL;
+
+    VkInstanceCreateInfo createInfo;
+    VkInstance instance;
+    uint32_t physDevCount;
+    VkPhysicalDevice physDevice;
+
+    static bool devPropertiesValid = false;
+    static VkPhysicalDeviceProperties devProperties;
+
+    if (devPropertiesValid) {
+        rval = &devProperties;
+        goto allDone;
+    }
+
+    // Save VK_DEVICE_LAYERS and VK_INSTANCE_LAYERS.
+    // We can't just save the pointer to the value, because getenv returns
+    // a pointer to where the value is stored, but we are going to change it.
+    char* pDL = vktrace_get_global_var("VK_DEVICE_LAYERS");
+    char* pIL = vktrace_get_global_var("VK_INSTANCE_LAYERS");
+    char *pSaveDL = NULL, *pSaveIL = NULL;
+    char* pEnvSave;
+    int lenDL = pDL ? strlen(pDL) : 0;
+    int lenIL = pIL ? strlen(pIL) : 0;
+    pEnvSave = vktrace_malloc(lenDL + lenIL + 2);
+    if (!pEnvSave) goto allDone;
+    if (pDL) {
+        pSaveDL = pEnvSave;
+        strncpy(pSaveDL, pDL, lenDL + 1);
+    }
+    if (pIL) {
+        pSaveIL = pEnvSave + lenDL + 1;
+        strncpy(pSaveIL, pIL, lenIL + 1);
+    }
+
+    // Set VK_DEVICE_LAYERS and VK_INSTANCE_LAYERS to null.
+    // We don't want any layers for this call to vKCreateInstance,
+    // especially the trace layer!
```

Comment:
On looking at this, I realized that this code is being run in the vktrace process, not the trace layer. The querying of the trace configuration should be done in the trace layer, because the vktrace command and the trace program can be on different systems. I'll have to change vktrace and the trace layer to create the file header inside the trace layer.